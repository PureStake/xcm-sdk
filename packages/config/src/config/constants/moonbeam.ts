import { createBalanceBuilder } from '../../balance';
import {
  Assets,
  Chain,
  MOONBEAM_ASSETS_CONFIGS as assets,
  MOONBEAM_CHINS_CONFIGS as chains,
  MoonChain,
  MOON_CHINS_CONFIGS,
} from '../../constants';
import {
  createExtrinsicBuilder,
  XTokensExtrinsicSuccessEvent,
} from '../../extrinsic';
import { MoonbeamAssets } from '../../interfaces';
import { createWithdrawBuilder } from '../../withdraw';
import { ChainXcmConfigs } from '../config.interfaces';

const config = MOON_CHINS_CONFIGS[MoonChain.Moonbeam];
const balance = createBalanceBuilder<MoonbeamAssets>();
const extrinsic = createExtrinsicBuilder<MoonbeamAssets>(config);
const withdraw = createWithdrawBuilder<MoonbeamAssets>();

export const MOONBEAM_CONFIGS: ChainXcmConfigs<MoonbeamAssets> = {
  [Assets.ACA]: {
    asset: assets[Assets.ACA],
    origin: chains[Chain.Acala],
    deposit: [
      {
        balance: balance.system(),
        extrinsic: extrinsic
          .xTokens()
          .transfer()
          .successEvent(XTokensExtrinsicSuccessEvent.TransferredMultiAssets)
          .params(chains[Chain.Acala], {
            Token: assets[Assets.ACA].originSymbol,
          }),
        origin: chains[Chain.Acala],
      },
    ],
    withdraw: [
      withdraw.xTokens({
        balance: balance.system(),
        destination: chains[Chain.Acala],
        existentialDeposit: 100_000_000_000,
        feePerWeight: 8,
      }),
    ],
  },
  [Assets.AUSD]: {
    asset: assets[Assets.AUSD],
    origin: chains[Chain.Acala],
    deposit: [],
    withdraw: [],
  },
  [Assets.DOT]: {
    asset: assets[Assets.DOT],
    origin: chains[Chain.Polkadot],
    deposit: [],
    withdraw: [],
  },
  [Assets.GLMR]: {
    asset: assets[Assets.GLMR],
    origin: config,
    deposit: [],
    withdraw: [],
  },
  [Assets.PARA]: {
    asset: assets[Assets.PARA],
    origin: chains[Chain.Parallel],
    deposit: [],
    withdraw: [],
  },
};
