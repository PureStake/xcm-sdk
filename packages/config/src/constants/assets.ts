export enum Asset {
  ACA = 'ACA',
  ASTR = 'ASTR',
  AUSD = 'AUSD',
  BNC = 'BNC',
  BSX = 'BSX',
  CRAB = 'CRAB',
  CRU = 'CRU',
  CSM = 'CSM',
  DEV = 'DEV',
  DOT = 'DOT',
  GLMR = 'GLMR',
  HKO = 'HKO',
  KAR = 'KAR',
  KBTC = 'KBTC',
  KINT = 'KINT',
  KMA = 'KMA',
  KSM = 'KSM',
  KUSD = 'KUSD',
  LIT = 'LIT',
  MOVR = 'MOVR',
  PARA = 'PARA',
  PARING = 'PARING',
  PHA = 'PHA',
  RMRK = 'RMRK',
  SDN = 'SDN',
  TEER = 'TEER',
  UNIT = 'UNIT',
  USDT = 'USDT',
}
