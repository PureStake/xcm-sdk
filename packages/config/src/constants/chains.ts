// eslint-disable-next-line import/no-cycle
import { MoonChain } from '../interfaces';

export enum MoonChainKey {
  MoonbaseAlpha = 'MoonbaseAlpha',
  Moonbeam = 'Moonbeam',
  Moonriver = 'Moonriver',
}
export const MOON_CHAINS: Record<MoonChainKey, MoonChain> = {
  [MoonChainKey.MoonbaseAlpha]: {
    key: MoonChainKey.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
    decimals: 18,
    unitsPerSecond: 50_000_000_000_000_000n,
  },
  [MoonChainKey.Moonbeam]: {
    key: MoonChainKey.Moonbeam,
    name: 'Moonbeam',
    ws: 'wss://wss.api.moonbeam.network',
    parachainId: 2004,
    decimals: 18,
    unitsPerSecond: 10_000_000_000_000_000_000n,
  },
  [MoonChainKey.Moonriver]: {
    key: MoonChainKey.Moonriver,
    name: 'Moonriver',
    ws: 'wss://wss.api.moonriver.moonbeam.network',
    parachainId: 2023,
    decimals: 18,
    unitsPerSecond: 100_000_000_000_000_000n,
  },
};

export enum ChainKey {
  Acala = 'Acala',
  AlphanetRelay = 'AlphanetRelay',
  Astar = 'Astar',
  AstarAlphanet = 'AstarAlphanet',
  BasiliskAlphanet = 'BasiliskAlphanet',
  Bifrost = 'Bifrost',
  Calamari = 'Calamari',
  CloverAlphanet = 'CloverAlphanet',
  CrustShadow = 'CrustShadow',
  CrustShadowAlphanet = 'CrustShadowAlphanet',
  Darwinia = 'Darwinia',
  DarwiniaAlphanet = 'DarwiniaAlphanet',
  Integritee = 'Integritee',
  IntegriteeAlphanet = 'IntegriteeAlphanet',
  InterBTCAlphanet = 'InterBTCAlphanet',
  Interlay = 'Interlay',
  Karura = 'Karura',
  Khala = 'Khala',
  Kintsugi = 'Kintsugi',
  Kusama = 'Kusama',
  LitentryAlphanet = 'LitentryAlphanet',
  Litmus = 'Litmus',
  Parallel = 'Parallel',
  ParallelHeiko = 'ParallelHeiko',
  Phala = 'Phala',
  Polkadot = 'Polkadot',
  Shiden = 'Shiden',
  Statemine = 'Statemine',
  StatemineAlphanet = 'StatemineAlphanet',
}
