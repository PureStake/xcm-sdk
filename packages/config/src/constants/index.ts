/* eslint-disable import/no-cycle */
export * from './assets';
export * from './chains';
