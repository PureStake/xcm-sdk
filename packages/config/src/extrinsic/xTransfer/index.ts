export * from './xTransfer';
export * from './xTransfer.constants';
export * from './xTransfer.interfaces';
