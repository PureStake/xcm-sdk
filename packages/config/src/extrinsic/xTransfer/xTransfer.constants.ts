export enum XTransferExtrinsic {
  Transfer = 'transfer',
}

export enum XTransferExtrinsicSuccessEvent {
  Deposited = 'Deposited',
  Withdrawn = 'Withdrawn',
}
