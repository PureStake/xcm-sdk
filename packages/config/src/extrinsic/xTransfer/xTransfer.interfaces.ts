import { Parents, XcmExtrinsicGetParams } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  XTransferExtrinsic,
  XTransferExtrinsicSuccessEvent,
} from './xTransfer.constants';

export interface XTransferPallet {
  pallet: ExtrinsicPallet.XTransfer;
  extrinsic: XTransferExtrinsic;
  successEvent: XTransferExtrinsicSuccessEvent;
  getParams: (params: XcmExtrinsicGetParams) => XTransferPalletParams;
}

export type XTransferPalletParams = [
  /**
   * asset
   */
  {
    id: {
      Concrete: XTransferConcreteParam;
    };
    fun: {
      Fungible: bigint;
    };
  },
  /**
   * destination
   */
  {
    parents: 1;
    interior: {
      X2: [
        {
          Parachain: number;
        },
        {
          AccountKey20: {
            /**
             * account
             */
            key: string;
          };
        },
      ];
    };
  },
  /**
   * weight
   */
  {
    refTime: 5_000_000_000;
    proofSize: 0;
  },
];

export interface XTransferConcreteParam {
  parents: Parents;
  interior: XTransferInteriorParam;
}

export type XTransferInteriorParam =
  | 'Here'
  | {
      X2: [
        {
          Parachain: number;
        },
        {
          PalletInstance: number;
        },
      ];
    };
