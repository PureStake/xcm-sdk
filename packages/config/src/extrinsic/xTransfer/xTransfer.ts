import { ChainConfig, MoonChainConfig } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  XTransferExtrinsic,
  XTransferExtrinsicSuccessEvent,
} from './xTransfer.constants';
import {
  XTransferConcreteParam,
  XTransferPallet,
} from './xTransfer.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function xTransfer(chain: MoonChainConfig) {
  return {
    transfer: () => transfer(chain),
  };
}

function transfer(chain: MoonChainConfig) {
  return {
    successEvent: (event: XTransferExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => {
        const createExtrinsic = getCreateExtrinsic(
          XTransferExtrinsic.Transfer,
          event,
          chain,
          origin,
        );

        return {
          here: () =>
            createExtrinsic(() => ({
              parents: 0,
              interior: 'Here',
            })),
          X2: (palletInstance: number) =>
            createExtrinsic(() => ({
              parents: 1,
              interior: {
                X2: [
                  {
                    Parachain: chain.parachainId,
                  },
                  {
                    PalletInstance: palletInstance,
                  },
                ],
              },
            })),
        };
      },
    }),
  };
}

function getCreateExtrinsic(
  extrinsic: XTransferExtrinsic,
  event: XTransferExtrinsicSuccessEvent,
  chain: MoonChainConfig,
  origin: ChainConfig,
) {
  return (getConcrete: () => XTransferConcreteParam): XTransferPallet => ({
    pallet: ExtrinsicPallet.XTransfer,
    extrinsic,
    successEvent: event,
    getParams: (account, amount) => [
      {
        id: {
          Concrete: getConcrete(),
        },
        fun: {
          Fungible: amount,
        },
      },
      {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: chain.parachainId,
            },
            {
              AccountKey20: {
                network: 'Any',
                key: account,
              },
            },
          ],
        },
      },
      origin.weight,
    ],
  });
}
