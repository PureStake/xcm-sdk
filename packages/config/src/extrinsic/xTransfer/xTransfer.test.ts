import { xTransfer } from './xTransfer';
import { Chain, MoonChain } from '../../constants';
import { XTransferExtrinsicSuccessEvent } from './xTransfer.constants';

describe('xTransfer', () => {
  const amount = 1000n;
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const chain = {
    chain: MoonChain.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
  };
  const origin = {
    chain: Chain.AlphanetRelay,
    name: 'Alphanet Relay',
    ws: 'wss://frag-moonbase-relay-rpc-ws.g.moonbase.moonbeam.network',
    weight: 1_000_000_000,
    parachainId: 0,
  };
  const extrinsic = xTransfer(chain);

  describe('transfer', () => {
    describe('here', () => {
      const cfg = extrinsic
        .transfer()
        .successEvent(XTransferExtrinsicSuccessEvent.Deposited)
        .origin(origin)
        .here();

      it('should be correct config', () => {
        expect(cfg).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(cfg.getParams(account, amount)).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      const cfg = extrinsic
        .transfer()
        .successEvent(XTransferExtrinsicSuccessEvent.Deposited)
        .origin(origin)
        .X2(99);

      it('should be correct config', () => {
        expect(cfg).toMatchSnapshot();
      });

      it('should get correct params', () => {
        expect(cfg.getParams(account, amount)).toMatchSnapshot();
      });
    });
  });
});
