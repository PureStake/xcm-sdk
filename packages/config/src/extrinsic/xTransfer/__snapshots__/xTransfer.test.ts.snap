// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xTransfer transfer here should be correct config 1`] = `
{
  "extrinsic": "transfer",
  "getParams": [Function],
  "pallet": "xTransfer",
  "successEvent": "Deposited",
}
`;

exports[`xTransfer transfer here should get correct params 1`] = `
[
  {
    "fun": {
      "Fungible": 1000n,
    },
    "id": {
      "Concrete": {
        "interior": "Here",
        "parents": 0,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;

exports[`xTransfer transfer x2 should be correct config 1`] = `
{
  "extrinsic": "transfer",
  "getParams": [Function],
  "pallet": "xTransfer",
  "successEvent": "Deposited",
}
`;

exports[`xTransfer transfer x2 should get correct params 1`] = `
[
  {
    "fun": {
      "Fungible": 1000n,
    },
    "id": {
      "Concrete": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "PalletInstance": 99,
            },
          ],
        },
        "parents": 1,
      },
    },
  },
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      ],
    },
    "parents": 1,
  },
  {
    "proofSize": 0,
    "refTime": 5000000000,
  },
]
`;
